framework/include/executors/Executor.h

```diff
@@ -0,0 +1,166 @@
+    /// descendant's results.  If success_msg is true, then all result output
+    /// that contains a message will be printed even if it converged/passed.
+    /// Otherwise, messages will only be printed for unconverged/failed results.
+    std::string
+    str(bool success_msg = false, const std::string & indent = "", const std::string & subname = "")
+    {
+      std::string s = indent + label(success_msg, subname) + "\n";
+      for (auto & entry : subs)
+        s += entry.second.str(success_msg, indent + "    ", entry.first);
+      return s;
+    }
+
+    /// Marks the result as passing/converged with the given msg text
+    /// describing detail about how things ran.  A result object is in this
+    /// state by default, so it is not necessary to call this function for
+    /// converged/passing scenarios.
+    void pass(const std::string & msg, bool overwrite = false)
+    {
+      mooseAssert(converged || overwrite,
+                  "cannot override nonconverged executioner result with a passing one");
+      ((void)(overwrite)); // avoid unused error due to assert
+      reason = msg;
+      converged = true;
+    }
+
+    /// Marks the result as failing/unconverged with the given msg text
+    /// describing detail about how things ran.
+    void fail(const std::string & msg)
+    {
+      reason = msg;
+      converged = false;
+    }
+
+    /// Records results from sub/internal executors in a executor's result.  When
+    /// child-executors return a result object following their execution, this
+    /// function should be called to add that info into the result heirarchy.
+    /// If the child executor was identified by a label/text from the input file
+    /// (e.g. via sub_solve1=foo_executor) - then "name" should be "sub_solve1".
+    bool record(const std::string & name, const Result & r)
+    {
+      subs[name] = r;
+      return r.convergedAll();
+    }
+
+    /// Returns false if any single executor in the current hierarchy of results
+    /// (i.e. including all child results accumulated recursively via record)
+    /// had a faild/unconverged return state.  Returns true otherwise.  This
+    /// is how convergence should generally be checked/tracked by executors -
+    /// rather than accessing e.g. the converged member directly.
+    bool convergedAll() const
+    {
+      if (!converged)
+        return false;
+      for (auto & entry : subs)
+        if (!entry.second.convergedAll())
+          return false;
+      return true;
+    }
+
+  private:
+    std::string label(bool success_msg, const std::string & subname = "")
+    {
+      std::string state_str =
+          success_msg || !converged ? (std::string("(") + (converged ? "pass" : "FAIL") + ")") : "";
+      return subname + (subname.empty() ? "" : ":") + _name + state_str +
+             ((success_msg || !converged) && !reason.empty() ? ": " + reason : "");
+    }
+    std::string _name;
+  };
```

Comment:
is there a way to mark that this is the end of Result ?
I had a double take when I saw the constructor below

Reply:
We don't have any conventions.  I'm inclined to leave it.  Just one of the joys of C++ syntax - nested namespace/class scopes.